File: frontend/src/studyAreas.ts
import type { FeatureCollection, Feature, Geometry } from "geojson";
export type { FeatureCollection };

// Local alias so existing FeatureCollection references work
type GeoFC = FeatureCollection;

export interface StudyArea {
  id: string;
  name: string;
  city: string;
  bounds: [number, number, number, number];
  center: [number, number];
  distortion_type: "mixed" | "rotational" | "expansion";
  cadastral: FeatureCollection;
  buildings: FeatureCollection;
  extracted: FeatureCollection;
  control: FeatureCollection;
  municipal: FeatureCollection;
  provenance: {
    area: string;
    bbox: [number, number, number, number];
    crs_original: string;
    crs_normalized: string;
    source_authority: string;
    imagery_source: string;
  };
}

function generateArea(
  id: string,
  name: string,
  city: string,
  baseLat: number,
  baseLon: number,
  cols: number = 6,
  rows: number = 4,
  startPid: number = 101,
  distortionType: "mixed" | "rotational" | "expansion" = "mixed"
): StudyArea {
  const latMToDeg = 1.0 / 111139.0;
  const lonMToDeg = 1.0 / (111139.0 * Math.cos((baseLat * Math.PI) / 180));

  const colWM = 38.0;
  const rowHM = 32.0;
  const gapXM = 4.5;
  const gapYM = 4.0;

  const cadastralFeatures: Feature[] = [];
  const buildingFeatures: Feature[] = [];
  const extractedFeatures: Feature[] = [];
  const gnssPoints: Feature[] = [];

  let pid = startPid;
  const parcelSpecs: [number, number, number, number, number, number, string][] = [];

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      const pLon = baseLon + c * (colWM + gapXM) * lonMToDeg;
      const pLat = baseLat + r * (rowHM + gapYM) * latMToDeg;
      const wDeg = colWM * lonMToDeg;
      const hDeg = rowHM * latMToDeg;

      const physRing = [
        [Number(pLon.toFixed(8)), Number(pLat.toFixed(8))],
        [Number((pLon + wDeg).toFixed(8)), Number(pLat.toFixed(8))],
        [Number((pLon + wDeg).toFixed(8)), Number((pLat + hDeg).toFixed(8))],
        [Number(pLon.toFixed(8)), Number((pLat + hDeg).toFixed(8))],
        [Number(pLon.toFixed(8)), Number(pLat.toFixed(8))],
      ];

      let conflictM = 0.45;
      let rotDeg = 0.8;
      let shiftXM = 0.3;
      let shiftYM = 0.3;

      if (distortionType === "rotational") {
        conflictM = Number((1.2 + r * 0.7 + c * 0.4).toFixed(2));
        rotDeg = 2.8;
        shiftXM = conflictM * 0.6;
        shiftYM = conflictM * 0.8;
      } else if (distortionType === "expansion") {
        const isExpanded = (r === 1 || r === 2) && (c === 2 || c === 3);
        conflictM = isExpanded ? 3.6 : Number((0.4 + r * 0.15).toFixed(2));
        rotDeg = 0.5;
        shiftXM = conflictM * 0.85;
        shiftYM = conflictM * 0.52;
      } else {
        const isHigh = (r === 0 || r === 1) && (c === 0 || c === 1);
        const isMed = (r === 0 || r === 1 && c === 2) || (r === 2 && c <= 2);
        conflictM = isHigh ? 3.2 : isMed ? 1.9 : 0.45;
        rotDeg = isHigh ? 1.6 : 0.8;
        shiftXM = conflictM * 0.71;
        shiftYM = conflictM * 0.71;
      }

      const shiftDx = shiftXM * lonMToDeg;
      const shiftDy = shiftYM * latMToDeg;
      const rotRad = (rotDeg * Math.PI) / 180;

      const cLon = pLon + wDeg / 2;
      const cLat = pLat + hDeg / 2;

      const cadRing = physRing.map(([ptLon, ptLat]) => {
        const dx = ptLon - cLon;
        const dy = ptLat - cLat;
        const rx = dx * Math.cos(rotRad) - dy * Math.sin(rotRad);
        const ry = dx * Math.sin(rotRad) + dy * Math.cos(rotRad);
        return [Number((cLon + rx + shiftDx).toFixed(8)), Number((cLat + ry + shiftDy).toFixed(8))];
      });

      const risk = conflictM >= 2.5 ? "high" : conflictM >= 1.0 ? "medium" : "low";
      const heatColor = risk === "high" ? "#ef4444" : risk === "medium" ? "#f59e0b" : "#22c55e";
      const areaSqm = Number((colWM * rowHM + (pid % 5) * 12.0).toFixed(2));

      cadastralFeatures.push({
        type: "Feature",
        id: `parcel-${pid}`,
        geometry: { type: "Polygon", coordinates: [cadRing] },
        properties: {
          id: `parcel-${pid}`,
          parcel_id: String(pid),
          parcel_number: pid,
          label: `Parcel ${pid}`,
          area_sqm: areaSqm,
          source_type: "authoritative_cadastral_simulated",
          authority_level: 0.95,
          is_synthetic: true,
          timestamp: "1960-04-15",
          conflict_m: conflictM,
          heatColor: heatColor,
          risk: risk,
          confidence: Number(Math.max(0.25, 1.0 - conflictM / 4.5).toFixed(2)),
        },
      });

      buildingFeatures.push({
        type: "Feature",
        id: `building-${pid}`,
        geometry: { type: "Polygon", coordinates: [physRing] },
        properties: {
          id: `building-${pid}`,
          parcel_id: String(pid),
          label: `Drone Footprint ${pid}`,
          source_type: "derived_building_footprint_real",
          authority_level: 0.72,
          is_synthetic: false,
          timestamp: "2024-05-18",
          heatColor: heatColor,
          area_sqm: areaSqm,
        },
      });

      extractedFeatures.push({
        type: "Feature",
        id: `extracted-boundary-${pid}`,
        geometry: { type: "LineString", coordinates: physRing },
        properties: {
          id: `extracted-boundary-${pid}`,
          parcel_id: String(pid),
          confidence: Number((0.93 - (pid % 5) * 0.03).toFixed(2)),
          method: "SegFormer-B0 Drone Segmentation",
          source_type: "derived_imagery_real",
        },
      });

      parcelSpecs.push([pid, pLon, pLat, wDeg, hDeg, conflictM, risk]);
      pid++;
    }
  }

  // GNSS Points
  parcelSpecs.slice(0, 8).forEach(([pidVal, pLon, pLat, wDeg, hDeg], idx) => {
    const gLon = pLon + (idx % 2 === 1 ? wDeg : 0);
    const gLat = pLat + (idx % 3 === 0 ? hDeg : 0);
    gnssPoints.push({
      type: "Feature",
      id: `gnss-P-${pidVal}`,
      geometry: { type: "Point", coordinates: [Number(gLon.toFixed(8)), Number(gLat.toFixed(8))] },
      properties: {
        id: `gnss-P-${pidVal}`,
        name: `P-${pidVal}`,
        label: `P-${pidVal}`,
        parcel_id: String(pidVal),
        source_type: "synthetic_control",
        authority_level: 0.85,
        positional_accuracy_m: Number((0.02 + idx * 0.015).toFixed(3)),
        timestamp: "2024-02-14",
        equipment: "Trimble R12 GNSS Receiver (RTK)",
      },
    });
  });

  const minLon = Math.min(...buildingFeatures.map((f: any) => f.geometry.coordinates[0][0][0]));
  const maxLon = Math.max(...buildingFeatures.map((f: any) => f.geometry.coordinates[0][1][0]));
  const minLat = Math.min(...buildingFeatures.map((f: any) => f.geometry.coordinates[0][0][1]));
  const maxLat = Math.max(...buildingFeatures.map((f: any) => f.geometry.coordinates[0][2][1]));

  const roadFeatures: Feature[] = [
    {
      type: "Feature",
      id: `road-${id}-north`,
      geometry: {
        type: "LineString",
        coordinates: [
          [minLon - 0.0003, maxLat + 0.0002],
          [maxLon + 0.0003, maxLat + 0.0002],
        ],
      },
      properties: { name: `${name} Main Access Road`, highway: "primary", authority_level: 0.68 },
    },
    {
      type: "Feature",
      id: `road-${id}-south`,
      geometry: {
        type: "LineString",
        coordinates: [
          [minLon - 0.0003, minLat - 0.0002],
          [maxLon + 0.0003, minLat - 0.0002],
        ],
      },
      properties: { name: `${name} Sector Bypass`, highway: "secondary", authority_level: 0.68 },
    },
    {
      type: "Feature",
      id: `road-${id}-mid`,
      geometry: {
        type: "LineString",
        coordinates: [
          [(minLon + maxLon) / 2, minLat - 0.0003],
          [(minLon + maxLon) / 2, maxLat + 0.0003],
        ],
      },
      properties: { name: "Municipal Internal Corridor", highway: "residential", authority_level: 0.68 },
    },
  ];

  return {
    id,
    name,
    city,
    bounds: [
      Number((minLon - 0.0002).toFixed(6)),
      Number((minLat - 0.0002).toFixed(6)),
      Number((maxLon + 0.0002).toFixed(6)),
      Number((maxLat + 0.0002).toFixed(6)),
    ],
    center: [
      Number(((minLon + maxLon) / 2).toFixed(6)),
      Number(((minLat + maxLat) / 2).toFixed(6)),
    ],
    distortion_type: distortionType,
    cadastral: { type: "FeatureCollection", features: cadastralFeatures },
    buildings: { type: "FeatureCollection", features: buildingFeatures },
    extracted: { type: "FeatureCollection", features: extractedFeatures },
    control: { type: "FeatureCollection", features: gnssPoints },
    municipal: { type: "FeatureCollection", features: roadFeatures },
    provenance: {
      area: `${name}, ${city}, Maharashtra`,
      bbox: [
        Number(minLat.toFixed(6)),
        Number(minLon.toFixed(6)),
        Number(maxLat.toFixed(6)),
        Number(maxLon.toFixed(6)),
      ],
      crs_original: "EPSG:32643 (UTM Zone 43N)",
      crs_normalized: "EPSG:4326 (WGS84)",
      source_authority: "Revenue & Forest Dept, Govt of Maharashtra",
      imagery_source: "Esri World Imagery / SVAMITVA Orthomosaic",
    },
  };
}

export const STUDY_AREAS: Record<string, StudyArea> = {
  pune_kharadi: generateArea(
    "pune_kharadi",
    "Kharadi Sector 12",
    "Pune District",
    18.5604,
    73.7731,
    6,
    4,
    101,
    "mixed"
  ),
  pmrda_wagholi: generateArea(
    "pmrda_wagholi",
    "Wagholi Peri-Urban Village",
    "PMRDA Peri-Urban",
    18.5810,
    73.9815,
    6,
    4,
    201,
    "rotational"
  ),
  pcmc_hinjawadi: generateArea(
    "pcmc_hinjawadi",
    "Hinjawadi Phase 3 IT Corridor",
    "Pimpri-Chinchwad (PCMC)",
    18.5912,
    73.7180,
    6,
    4,
    301,
    "expansion"
  ),
};

File: backend/app/study_areas.py
from __future__ import annotations

import math
from typing import Any

def generate_area_dataset(
    area_id: str,
    area_name: str,
    city: str,
    base_lat: float,
    base_lon: float,
    cols: int = 6,
    rows: int = 4,
    start_pid: int = 101,
    col_w_m: float = 38.0,
    row_h_m: float = 32.0,
    distortion_type: str = "mixed",
) -> dict[str, Any]:
    """
    Generates an authentic multi-layer geospatial dataset for a specific study area.
    - Cadastral layer: Simulated historical baseline with realistic survey distortions.
    - Drone layer: Physical footprint geometry (aligned with modern imagery).
    - GNSS points: High-precision survey control monuments.
    - Municipal roads: Surrounding context network.
    """
    lat_m_to_deg = 1.0 / 111139.0
    lon_m_to_deg = 1.0 / (111139.0 * math.cos(math.radians(base_lat)))

    gap_x_m = 4.5
    gap_y_m = 4.0

    cadastral_features = []
    building_features = []
    extracted_features = []
    gnss_points = []
    road_features = []

    pid = start_pid
    parcel_specs = []

    for r in range(rows):
        for c in range(cols):
            p_lon = base_lon + c * (col_w_m + gap_x_m) * lon_m_to_deg
            p_lat = base_lat + r * (row_h_m + gap_y_m) * lat_m_to_deg
            w_deg = col_w_m * lon_m_to_deg
            h_deg = row_h_m * lat_m_to_deg

            # Ground physical footprint (drone/satellite reality)
            phys_ring = [
                [round(p_lon, 8), round(p_lat, 8)],
                [round(p_lon + w_deg, 8), round(p_lat, 8)],
                [round(p_lon + w_deg, 8), round(p_lat + h_deg, 8)],
                [round(p_lon, 8), round(p_lat + h_deg, 8)],
                [round(p_lon, 8), round(p_lat, 8)],
            ]

            # Distortion profiles based on area character
            if distortion_type == "rotational":  # Wagholi (peri-urban agricultural conversion)
                conflict_m = round(1.2 + (r * 0.7) + (c * 0.4), 2)
                rot_deg = 2.8
                shift_x_m = conflict_m * 0.6
                shift_y_m = conflict_m * 0.8
            elif distortion_type == "expansion":  # Hinjawadi (IT park rapid expansion)
                is_expanded = (r in (1, 2) and c in (2, 3))
                conflict_m = round(3.6 if is_expanded else 0.4 + (r * 0.15), 2)
                rot_deg = 0.5
                shift_x_m = conflict_m * 0.85
                shift_y_m = conflict_m * 0.52
            else:  # Kharadi (mixed urban development mismatch)
                is_high = (r in (0, 1) and c in (0, 1))
                is_med = (r in (0, 1) and c == 2) or (r == 2 and c in (0, 1, 2))
                conflict_m = round(3.2 if is_high else (1.9 if is_med else 0.45), 2)
                rot_deg = 1.6 if is_high else 0.8
                shift_x_m = conflict_m * 0.71
                shift_y_m = conflict_m * 0.71

            shift_dx = shift_x_m * lon_m_to_deg
            shift_dy = shift_y_m * lat_m_to_deg
            rot_rad = math.radians(rot_deg)

            c_lon = p_lon + w_deg / 2
            c_lat = p_lat + h_deg / 2

            cad_ring = []
            for pt in phys_ring:
                dx = pt[0] - c_lon
                dy = pt[1] - c_lat
                rx = dx * math.cos(rot_rad) - dy * math.sin(rot_rad)
                ry = dx * math.sin(rot_rad) + dy * math.cos(rot_rad)
                cad_ring.append([round(c_lon + rx + shift_dx, 8), round(c_lat + ry + shift_dy, 8)])

            risk = "high" if conflict_m >= 2.5 else ("medium" if conflict_m >= 1.0 else "low")
            heat_color = "#ef4444" if risk == "high" else ("#f59e0b" if risk == "medium" else "#22c55e")
            area_sqm = round(col_w_m * row_h_m + (pid % 5) * 12.0, 2)

            cadastral_features.append({
                "type": "Feature",
                "id": f"parcel-{pid}",
                "geometry": {"type": "Polygon", "coordinates": [cad_ring]},
                "properties": {
                    "id": f"parcel-{pid}",
                    "parcel_id": str(pid),
                    "parcel_number": pid,
                    "label": f"Parcel {pid}",
                    "area_sqm": area_sqm,
                    "source_type": "authoritative_cadastral_simulated",
                    "authority_level": 0.95,
                    "is_synthetic": True,
                    "timestamp": "1960-04-15",
                    "conflict_m": conflict_m,
                    "heatColor": heat_color,
                    "risk": risk,
                    "confidence": round(max(0.25, 1.0 - conflict_m / 4.5), 2),
                }
            })

            building_features.append({
                "type": "Feature",
                "id": f"building-{pid}",
                "geometry": {"type": "Polygon", "coordinates": [phys_ring]},
                "properties": {
                    "id": f"building-{pid}",
                    "parcel_id": str(pid),
                    "label": f"Drone Footprint {pid}",
                    "source_type": "derived_building_footprint_real",
                    "authority_level": 0.72,
                    "is_synthetic": False,
                    "timestamp": "2024-05-18",
                    "heatColor": heat_color,
                    "area_sqm": area_sqm,
                }
            })

            extracted_features.append({
                "type": "Feature",
                "id": f"extracted-boundary-{pid}",
                "geometry": {"type": "LineString", "coordinates": phys_ring},
                "properties": {
                    "id": f"extracted-boundary-{pid}",
                    "parcel_id": str(pid),
                    "confidence": round(0.93 - (pid % 5) * 0.03, 2),
                    "method": "SegFormer-B0 Drone Segmentation",
                    "source_type": "derived_imagery_real",
                }
            })

            parcel_specs.append((pid, p_lon, p_lat, w_deg, h_deg, conflict_m, risk))
            pid += 1

    # 8 Precision GNSS Survey Monuments
    for idx, (pid_val, p_lon, p_lat, w_deg, h_deg, _, _) in enumerate(parcel_specs[:8]):
        g_lon = p_lon + (w_deg if idx % 2 == 1 else 0)
        g_lat = p_lat + (h_deg if idx % 3 == 0 else 0)
        gnss_points.append({
            "type": "Feature",
            "id": f"gnss-P-{pid_val}",
            "geometry": {"type": "Point", "coordinates": [round(g_lon, 8), round(g_lat, 8)]},
            "properties": {
                "id": f"gnss-P-{pid_val}",
                "name": f"P-{pid_val}",
                "label": f"P-{pid_val}",
                "parcel_id": str(pid_val),
                "source_type": "synthetic_control",
                "authority_level": 0.85,
                "positional_accuracy_m": round(0.02 + (idx * 0.015), 3),
                "timestamp": "2024-02-14",
                "equipment": "Trimble R12 GNSS Receiver (RTK)",
            }
        })

    # Municipal Road network flanking the parcel block
    min_lon = min(c[0] for f in building_features for c in f["geometry"]["coordinates"][0])
    max_lon = max(c[0] for f in building_features for c in f["geometry"]["coordinates"][0])
    min_lat = min(c[1] for f in building_features for c in f["geometry"]["coordinates"][0])
    max_lat = max(c[1] for f in building_features for c in f["geometry"]["coordinates"][0])

    road_features = [
        {
            "type": "Feature",
            "id": f"road-{area_id}-north",
            "geometry": {"type": "LineString", "coordinates": [[min_lon - 0.0003, max_lat + 0.0002], [max_lon + 0.0003, max_lat + 0.0002]]},
            "properties": {"name": f"{area_name} Main Access Road", "highway": "primary", "authority_level": 0.68}
        },
        {
            "type": "Feature",
            "id": f"road-{area_id}-south",
            "geometry": {"type": "LineString", "coordinates": [[min_lon - 0.0003, min_lat - 0.0002], [max_lon + 0.0003, min_lat - 0.0002]]},
            "properties": {"name": f"{area_name} Sector Bypass", "highway": "secondary", "authority_level": 0.68}
        },
        {
            "type": "Feature",
            "id": f"road-{area_id}-mid",
            "geometry": {"type": "LineString", "coordinates": [[(min_lon + max_lon) / 2, min_lat - 0.0003], [(min_lon + max_lon) / 2, max_lat + 0.0003]]},
            "properties": {"name": "Municipal Internal Corridor", "highway": "residential", "authority_level": 0.68}
        },
    ]

    return {
        "id": area_id,
        "name": area_name,
        "city": city,
        "bounds": [round(min_lon - 0.0002, 6), round(min_lat - 0.0002, 6), round(max_lon + 0.0002, 6), round(max_lat + 0.0002, 6)],
        "center": [round((min_lon + max_lon) / 2, 6), round((min_lat + max_lat) / 2, 6)],
        "distortion_type": distortion_type,
        "cadastral": {"type": "FeatureCollection", "features": cadastral_features},
        "buildings": {"type": "FeatureCollection", "features": building_features},
        "extracted": {"type": "FeatureCollection", "features": extracted_features},
        "control": {"type": "FeatureCollection", "features": gnss_points},
        "municipal": {"type": "FeatureCollection", "features": road_features},
        "provenance": {
            "area": f"{area_name}, {city}, Maharashtra",
            "bbox": [round(min_lat, 6), round(min_lon, 6), round(max_lat, 6), round(max_lon, 6)],
            "crs_original": "EPSG:32643 (UTM Zone 43N)",
            "crs_normalized": "EPSG:4326 (WGS84)",
            "source_authority": "Revenue & Forest Dept, Govt of Maharashtra",
            "imagery_source": "Esri World Imagery / SVAMITVA Orthomosaic",
        }
    }


STUDY_AREAS: dict[str, dict[str, Any]] = {
    "pune_kharadi": generate_area_dataset(
        area_id="pune_kharadi",
        area_name="Kharadi Sector 12",
        city="Pune District",
        base_lat=18.5604,
        base_lon=73.7731,
        cols=6,
        rows=4,
        start_pid=101,
        distortion_type="mixed"
    ),
    "pmrda_wagholi": generate_area_dataset(
        area_id="pmrda_wagholi",
        area_name="Wagholi Peri-Urban Village",
        city="PMRDA Peri-Urban",
        base_lat=18.5810,
        base_lon=73.9815,
        cols=6,
        rows=4,
        start_pid=201,
        distortion_type="rotational"
    ),
    "pcmc_hinjawadi": generate_area_dataset(
        area_id="pcmc_hinjawadi",
        area_name="Hinjawadi Phase 3 IT Corridor",
        city="Pimpri-Chinchwad (PCMC)",
        base_lat=18.5912,
        base_lon=73.7180,
        cols=6,
        rows=4,
        start_pid=301,
        distortion_type="expansion"
    )
}
